### DISCUSSION TITLE
Upgrade TWAP to poll data based on a controller

### DISCUSSION THREAD

**Discussion by Draeieg** [pavlovcik/uad-ui#340] [General]
The twap price we will target is an aggregation of various data sources grouped depending on their gas cost and origin, current groups are

ChainLink price feeds
Uma optimistic oracles (prob will get axed)
Median aggregation of dexes
 to make various sources

These data sources will be called based on market conditions. When there is not a lot of volume amount of transactions is minimum and volatility is low only the cheapest price feeds will be called, as the conditions change to a harsher, more lively market data sources will progressively be added until the twap becomes an aggregation of all possible data sources

This has the purpose of making our oracle calls as efficient as possible by reducing the gas used determining the twap by reducing our set of data sources in “safer” market conditions while at the same time being able to ramp the robustness during higher volatility periods 

These calls will be determined by a proportional controller (this ideally will evolve into full PID in future versions) based on 
Volume per unit of time (daily, hourly or block based)
Amount of tx per unit of time (daily, hourly or block based)

The controller will work based on 3 variables,  a simplified version of them is

1. Volume and tx aggregation (V), as the process variable
2. Oracle base polls (BP), as the set variable. which has 2 elements
    P is the actual amount of polls per unit of time 
    B is a baseline variable, its a  function that will increase or decrease overtime to change the net polls done and is affected by the 
     third variable generated at the output, it's the feedback element of the system
3. Poll rate(R), the output variable. is the result of the V-BP proportion, it will increase P variable (the actual amount of polls) as well as the delta of B 

The resulting equations will be then

V-BP=R, since the controller always wants to be at 0 and V is what where measuring then;
R=BP, as B is an independent function of time only P can be changed as a result of an R change

<img width="429" alt="Untitled Diagram drawio" src="https://user-images.githubusercontent.com/88983138/202809083-306e37eb-5d27-4895-b05b-b740ca875e2e.png">

In this simplified example when volume rises the difference between the base polls (BP) and volume (V) increases causing the poll rate (R) to increase, since the controller always wants to be at 0 the net polls (P) increase
As the difference goes back to 0 due to base polls increasing in tandem with volume the baseline (B) increases as an effect of the R increase allowing the net polls go down while the baseline stays the same, as this new volume is deemed “new normal”
When volume decreases P has to go down to bring the difference back to 0, as R is a negative number. This negative R causes B to also go down over time, increasing the net polls after given amount of time

Baseline function
This will change what the controller defines as “normal volume” , a simple way to see it is in an unit of time the controller wants it to be the equal to R

By example “it should go from 1 to 3 in a week”

Contract structure 

The current contract structure is based on the way I design electronics, this means it will probably be shit in some aspects, for a more exact spec look at RAI controller contracts 
an individual contract for each individual data source, in this version that means 4
Individual interfaces to normalize the data sources to a common format for the controller to read, another 4
A controller contract

This puts the total to 9 in this current iteration

**Comment by 0x4007** [comment]
# Minimum Viable Oracle (MVO)

Can you be specific on which oracles you have in mind to start off this system? For example 

"we need the following markets' data:

- Curve uAD-3pool
- Uniswap uAD-ETH

In order to create a minimum viable oracle with this design"

# Other Considerations

We discussed Credit redemption and how it could make the token inflationary at a variable rate, depending on the health of our economy. Perhaps this is out of scope for this TWAP oracle but I know you did some research on it and I'm not sure if we should have a separate discussion for it, or if we should include those considerations in here. 

Since I'm already talking about it here, I want to add that we should also look at the total circulating supply of Dollars, and distinguish which are in control of the protocol and which are not. If the Dollars are in control of other protocols/whales, then we should always plan for a worst case scenario of them dumping 100% and still being able to pay off those liabilities (given our current liquidity and/or price + collateral in the bank.)

This means that we should always be coasting along over-collateralized (unless for some reason, perhaps while bootstrapping things, that Ubiquity controls over half of the circulating supply of Dollars, then we could be under collateralized because we wouldn't dump on ourselves after everybody else dumped on us)

**Comment by Draeieg** [comment]
# Markets to track for the MVO

## Curve markets
1) *3pool (we do this already)*
2) median of the top 5 pools other than 3 pool by TVL 
	currently (Frax-3pool, Frax-USDC, MIM-3pool, alusd-3pool, susd-3pool)
3) Median of the top 5 pools other than 3pool by volume
4) median of all curve pools (hiding very small pools 

>these measures have the vulnerability of over reliance on 3pool, if the 3pool goes down so most if not all of curve measures

## Uniswap markets by TVL

1) Median of USDC markets
	1.1 USDC-DAI
	1.2 USDC-USDT
	1.3 USDC-FRAX
	1.4 USDC-BUSD
	1.5 USDC-TUSD
2) Median of USDT markets
	2.1 USDT-USDC
	2.2 USDT-FRAX
	2.3 USDT-DAI
	2.4 USDT-BUSD
	2.5 USDT-OUSD

3) Median of DAI markets 
	3.1 DAI-USDC
	3.2 DAI-FRAX
	3.3 DAI-USDT
	3.4 DAI-LUSD
	3.5 DAI-BUSD

4)  Median on FRAX markets
	4.1 FRAX-USDC
	4.2 FRAX-USDT
	4.3 FRAX-DAI
	4.4 FRAX-LUSD
	4.5 FRAX-SUSD
5) Median of mixed markets
	5.1 MIM-USDC
	5.2 TUSD-USDC
	5.3 OUSD-USDT

## Median of Chainlink Data Feeds

1) USDC/USD 0x8fffffd4afb6115b954bd326cbe7b4ba576818f6
2) USDT/USD 0x3e7d1eab13ad0104d2750b8863b489d65364e32d
3) DAI/USD 0xaed0c38402a5d19df6e4c03f4e2dced6e29c1ee9
4) FRAX/USD 0xb9e1e3a9feff48998e45fa90847ed4d467e8bcfd
5) sUSD/USD 0xad35bd71b9afe6e4bdc266b345c198eadef9ad94

These 5 will work as a baseline when we start to use data from chainlink, as volume and volatility grow more data feeds are added, similar to how we would take all usd pairs from curve 


Note for this baseline we wont be using our own price as a reference as that would result in unwanted feedback loops

**Reply by 0x4007** [reply]
> these measures have the vulnerability of over reliance on 3pool, if the 3pool goes down so most if not all of curve measures

We should be dynamically calculating the largest pools so you shouldn't need to cherry pick specific markets. Perhaps thats out of scope for v1 of the oracle but its something I hope to crack the code on soon.

**Comment by Draeieg** [comment]
# MVO spec v1

The initial controller will be a simple P controller, we should look into the possibility of forking RAI as the math should be the same (if not simpler)

variables are 
>V, for volume
>P, for oracle polls
>R, for rate  

the governing equation for the controller will be 

$v-p=r$

A controller always wants to be at 0, as such everytime volume goes up so will the amount of polls the oracle makes, this can be simplified as 

$v=p$

Our current volume will be considered baseline (1), ideally the volume variable will be logarithmic however for simplicity of v1 I have 2 options

1) A 20% increase represents 1 extra poll, so if we go from 1k daily volume to 1.2k we do 2 polls instead of 1
2) Each 10k USD or roughly each 10 ETH will represent an extra 1, so going from 10 eth daily volume to 20 eth will represent an increase from 1 to 2 

# Markets

The following list is all of the markets that v1 should track order depends on the gas cost of each source

 >A small bounty can be taken from this part and have someone see how much each one costs in order to develop the proper table 

## Curve markets
1) *3pool (we do this already)*
2) median of the top 5 pools other than 3 pool by TVL 
	currently (Frax-3pool, Frax-USDC, MIM-3pool, alusd-3pool, susd-3pool)
3) Median of the top 5 pools other than 3pool by volume
4) median of all curve pools (hiding very small pools 

>these measures have the vulnerability of over reliance on 3pool, if the 3pool goes down so most if not all of curve measures

## Uniswap markets by TVL

1) Median of USDC markets
	1.1 USDC-DAI
	1.2 USDC-USDT
	1.3 USDC-FRAX
	1.4 USDC-BUSD
	1.5 USDC-TUSD
2) Median of USDT markets
	2.1 USDT-USDC
	2.2 USDT-FRAX
	2.3 USDT-DAI
	2.4 USDT-BUSD
	2.5 USDT-OUSD

3) Median of DAI markets 
	3.1 DAI-USDC
	3.2 DAI-FRAX
	3.3 DAI-USDT
	3.4 DAI-LUSD
	3.5 DAI-BUSD

4)  Median on FRAX markets
	4.1 FRAX-USDC
	4.2 FRAX-USDT
	4.3 FRAX-DAI
	4.4 FRAX-LUSD
	4.5 FRAX-SUSD
5) Median of mixed markets
	5.1 MIM-USDC
	5.2 TUSD-USDC
	5.3 OUSD-USDT

## Median of Chainlink Data Feeds

1) USDC/USD 0x8fffffd4afb6115b954bd326cbe7b4ba576818f6
2) USDT/USD 0x3e7d1eab13ad0104d2750b8863b489d65364e32d
3) DAI/USD 0xaed0c38402a5d19df6e4c03f4e2dced6e29c1ee9
4) FRAX/USD 0xb9e1e3a9feff48998e45fa90847ed4d467e8bcfd
5) sUSD/USD 0xad35bd71b9afe6e4bdc266b345c198eadef9ad94

Since ChainLink already does medians on their oracles this section is probably overkill, but depending on gas cost it can be a nice to have

**Comment by Draeieg** [comment]
# Core concepts 

The twap price will be the median of various data sources grouped depending on their gas cost and origin,

These data sources will be called based on volume (in future versions also amount of tx and volatility) when volume is low only 1-2 of the cheapest price feeds will be called, as volume rises more data sources are polled and

This will make our twap as efficient as possible by reducing gas when there is not a lot going on while maintaining a robust infrastructure when shit goes down

These calls will be determined by a proportional controller (this ideally will evolve into full PID in future versions) based on Volume per Block

# Equations

1) Polls ( P ) is the amount of data sources we use for the TWAP, this will be related to an ordered list where by example:
 1=3pool/usd on curve
 2=usdc/usd on chainlink
 3=usdc/dai on uniswap
If there is an increase in P all previous data sources have to be aggregated to the median, this means 
if P is 1 then TWAP=3pool/usd
if P is 2 then TWAP= median of (3pool/usd, usdc/usd)
if P is 3 then TWAP= median of (3pool/usd, usdc/usd, usdc/dai)

	*P will be rounded up, so 0.5=1
2)  Volume (V), as the process variable. For simplicity our current volume will be considered 1 (baseline)
>A 20% increase represents 1 extra poll, so if we go from 1k daily volume  to    1.2k we do 2 polls instead of 1
		a Simple algorithm I propose for this is:  
		If $\large{\frac {CurrentVolume} {BaselineVolume}>1.2}$;  then
			$P=P+1$
			BaselineVolume=Current volume 
			however if someone comes up anything more efficient/better we use thta  
			instead
3) Baseline (B) this will be what makes this system viable long term as it will regulate the amount of data sources and determinate what our *new normal* is

	If BaselineVolume has remained constant for more than 24h (about 6646 eth blocks) withing a 2% margin of error then B will start to increase, as soon as Baselinevolume changes its value B will be reset back to 1
	
	This will be a value between 1 and 2 that will increase linearly in the span of 24h so $0.000150466445983$ per block


4) Finally we have the output from the System, R. we'll start to use the full controller equation as that is where R comes from

	$V-BP=R$

	The controller always wants to be at 0, in code this will look along the lines of
	>$V-BP=R$
	If $R \not = 0$, (withing a margin of error of 2%). and since B is back to 1
	$P=V$
	After the B function starts working 
	$P=\frac P B$


## Graphs/Simulations

<img width="429" alt="202809083-306e37eb-5d27-4895-b05b-b740ca875e2e" src="https://user-images.githubusercontent.com/88983138/211424196-4f5418e9-34b1-47f0-bc8c-a32814c544c5.png">


In this example when volume rises the difference between the base polls (BP) and volume (V) increases causing the poll rate (R) to increase, since the controller always wants to be at 0 the net polls (P) increase  
As the difference goes back to 0 the baseline (B) increases allowing the net polls go down while the line stays the same, as this new volume is deemed “new normal”  
When volume decreases P has to go down to bring the difference back to 0, as R is a negative number. This negative R causes B to also go down over time, increasing the net polls after given amount of time

An example of the functionality can be found [here](https://www.desmos.com/calculator/haa2c0famu) the general idea is move the value sliders around while trying to maintain the blue line at 0



# Markets

The following list is all of the markets that v1 should track order depends on the gas cost of each source

 >A small bounty can be taken from this part and have someone see how much each one costs in order to develop the proper table 

## Curve markets
1) *3pool (we do this already)*
2) median of the top 5 pools other than 3 pool by TVL 
	currently (Frax-3pool, Frax-USDC, MIM-3pool, alusd-3pool, susd-3pool)
3) Median of the top 5 pools other than 3pool by volume
4) median of all curve pools (hiding very small pools 

>these measures have the vulnerability of over reliance on 3pool, if the 3pool goes down so most if not all of curve measures

## Uniswap markets by TVL

1) Median of USDC markets
	1.1 USDC-DAI
	1.2 USDC-USDT
	1.3 USDC-FRAX
	1.4 USDC-BUSD
	1.5 USDC-TUSD
2) Median of USDT markets
	2.1 USDT-USDC
	2.2 USDT-FRAX
	2.3 USDT-DAI
	2.4 USDT-BUSD
	2.5 USDT-OUSD

3) Median of DAI markets 
	3.1 DAI-USDC
	3.2 DAI-FRAX
	3.3 DAI-USDT
	3.4 DAI-LUSD
	3.5 DAI-BUSD

4)  Median on FRAX markets
	4.1 FRAX-USDC
	4.2 FRAX-USDT
	4.3 FRAX-DAI
	4.4 FRAX-LUSD
	4.5 FRAX-SUSD
5) Median of mixed markets
	5.1 MIM-USDC
	5.2 TUSD-USDC
	5.3 OUSD-USDT

## Median of Chainlink Data Feeds

1) USDC/USD 0x8fffffd4afb6115b954bd326cbe7b4ba576818f6
2) USDT/USD 0x3e7d1eab13ad0104d2750b8863b489d65364e32d
3) DAI/USD 0xaed0c38402a5d19df6e4c03f4e2dced6e29c1ee9
4) FRAX/USD 0xb9e1e3a9feff48998e45fa90847ed4d467e8bcfd
5) sUSD/USD 0xad35bd71b9afe6e4bdc266b345c198eadef9ad94

Since ChainLink already does medians on their oracles this section is probably overkill, but depending on gas cost it can be a nice to have

**Reply by Draeieg** [reply]
@pavlovcik

**Reply by Draeieg** [reply]
![oracle diagram](https://user-images.githubusercontent.com/88983138/221085046-7f081b89-86af-4c24-b125-b20f2a00f657.png)


A spreadsheet with tests for different scenarios can be found [here](https://docs.google.com/spreadsheets/d/1Kcnf0pwPIBFg6W5b4QALsvfFUT92tithdb9bYB1ODro/edit#gid=1501261885)

**Reply by 0x4007** [reply]
This is a nice diagram thanks for making it.

**Comment by rndquu** [comment]
@Draeieg 

so the overall idea is to create an efficient TWAP based on:
- curve markets
- uniswap markets
- chainlink feeds

questions:
1. The whole TWAP structure look very similar to reflexer's RAI, right? 
2. Do you think intergrating uniswap markets is a good start? It seems to be the simplest task out of all.

**Reply by Draeieg** [reply]
1) yes this was inspired by RAI. plan is to fork RAI's controller but repurpose it for this 
2) I'm not a solidity dev so if you say that's the easiest then let's go with it.

Another good place to start IMO is with the array with all of the markets and ther gas cost
